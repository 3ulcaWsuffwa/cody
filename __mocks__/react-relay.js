const ReactRelay = require("react-relay");

module.exports = {
  ...ReactRelay,
  // Mock createFragmentContainer HOC so it just returns the inner Component
  createFragmentContainer: Component => Component
};
