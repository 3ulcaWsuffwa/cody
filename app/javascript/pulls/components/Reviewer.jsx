// @flow

import React from "react";
import { createFragmentContainer, graphql } from "react-relay";

function statusToOcticon(status: string) {
  switch (status) {
    case "pending_approval":
      return (
        <span className="icon">
          <i className="fa fa-square-o" />
        </span>
      );
    case "approved":
      return (
        <span className="icon">
          <i className="fa fa-check-square-o" />
        </span>
      );
    default:
      return status;
  }
}

const Reviewer = ({ reviewer }) =>
  <div className="level">
    <div className="level-left">
      <div className="level-item">
        {statusToOcticon(reviewer.status)}
      </div>
      <div className="level-item">
        <strong>
          {reviewer.login}
        </strong>
      </div>
      <div className="level-item">
        {reviewer.reviewRule != null ? reviewer.reviewRule.name : false}
      </div>
    </div>
  </div>;

export default createFragmentContainer(
  Reviewer,
  graphql`
    fragment Reviewer_reviewer on Reviewer {
      id
      login
      status
      reviewRule {
        name
      }
    }
  `
);
