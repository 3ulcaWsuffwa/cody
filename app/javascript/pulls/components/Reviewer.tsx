import React from "react";
import { Circle, CheckCircle } from "react-feather";
import { createFragmentContainer, graphql } from "react-relay";
import type {
  Reviewer_reviewer,
  ReviewerStatus,
} from "./__generated__/Reviewer_reviewer.graphql";

function statusToOcticon(status: ReviewerStatus) {
  switch (status) {
    case "APPROVED":
      return (
        <>
          <span className="sr-only">Approved</span>
          <CheckCircle className="flex-shrink-0 mr-1.5 h-5 w-5 text-green-500" />
        </>
      );
    case "PENDING_APPROVAL":
    default:
      return (
        <>
          <span className="sr-only">Pending approval</span>
          <Circle className="flex-shrink-0 mr-1.5 h-5 w-5 text-yellow-300" />
        </>
      );
  }
}

function Reviewer({ reviewer }: { reviewer: Reviewer_reviewer }): JSX.Element {
  return (
    <li>
      <div className="flex items-center px-4 py-4 sm:px-6">
        <div className="min-w-0 flex flex-1 items-center">
          <div className="flex-shrink-0">
            {statusToOcticon(reviewer.status)}
          </div>
          <div className="min-w-0 flex-1 px-4 md:grid md:grid-cols-2 md:gap-4">
            <div>
              <p className="text-sm font-medium truncate" data-testid="reviewer-login">{reviewer.login}</p>
              {reviewer.reviewRule != null ? <p className="text-sm truncate text-gray-500" data-testid="review-rule-name">{reviewer.reviewRule.name}</p> : false}
            </div>
          </div>
        </div>
      </div>
    </li>
  );
}

export default createFragmentContainer(Reviewer, {
  reviewer: graphql`
    fragment Reviewer_reviewer on Reviewer {
      id
      login
      status
      reviewRule {
        name
      }
    }
  `,
});
