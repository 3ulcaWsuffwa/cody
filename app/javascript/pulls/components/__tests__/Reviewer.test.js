import React from "react";
import { render, screen } from "@testing-library/react";
import { createMockEnvironment, MockPayloadGenerator } from "relay-test-utils";
import { QueryRenderer, graphql } from "react-relay";
import Reviewer from "../Reviewer";
import "@testing-library/jest-dom/extend-expect";

const TestComponent = ({ environment }) => (
  <QueryRenderer
    environment={environment}
    query={graphql`
      query ReviewerTestSnapshotQuery @relay_test_operation {
        viewer {
          repository(owner: "test", name: "test") {
            pullRequest(number: "42") {
              reviewers {
                edges {
                  node {
                    ...Reviewer_reviewer
                  }
                }
              }
            }
          }
        }
      }
    `}
    variables={{}}
    render={({ error, props }) => {
      if (error) {
        return error.message;
      } else if (props) {
        return (
          <Reviewer
            reviewer={
              props.viewer.repository.pullRequest.reviewers.edges[0].node
            }
          />
        );
      } else {
        return <div>Loading...</div>;
      }
    }}
  />
);

let environment;
beforeEach(() => {
  environment = createMockEnvironment();
});

test("Reviewer snapshot test", () => {
  let renderedComponent = render(<TestComponent environment={environment} />);

  environment.mock.resolveMostRecentOperation((operation) =>
    MockPayloadGenerator.generate(operation)
  );

  let fragment = renderedComponent.asFragment();
  expect(fragment).toMatchSnapshot();
});

test("Reviewer renders the reviewer's login", () => {
  render(<TestComponent environment={environment} />);

  environment.mock.resolveMostRecentOperation((operation) =>
    MockPayloadGenerator.generate(operation, {
      Reviewer(_, generatedId) {
        return {
          id: `test-id-${generatedId()}`,
          login: "aergonaut",
          status: "APPROVED",
          reviewRule: null,
        };
      },
    })
  );

  expect(screen.getByTestId("reviewer-login")).toHaveTextContent("aergonaut");
});

test("Reviewer renders the review rule name if it is given", () => {
  render(<TestComponent environment={environment} />);

  environment.mock.resolveMostRecentOperation((operation) =>
    MockPayloadGenerator.generate(operation, {
      Reviewer(_, generatedId) {
        return {
          id: `test-id-${generatedId()}`,
          login: "aergonaut",
          status: "APPROVED",
          reviewRule: {
            name: "Foobar",
          },
        };
      },
    })
  );

  expect(screen.getByTestId("review-rule-name")).toHaveTextContent("Foobar");
});

test("Reviewer renders Pending Approval icon", () => {
  render(<TestComponent environment={environment} />);

  environment.mock.resolveMostRecentOperation((operation) =>
    MockPayloadGenerator.generate(operation, {
      Reviewer(_, generatedId) {
        return {
          id: `test-id-${generatedId()}`,
          login: "aergonaut",
          status: "PENDING_APPROVAL",
          reviewRule: {
            name: "Foobar",
          },
        };
      },
    })
  );

  expect(screen.getByTitle("Pending approval")).toBeDefined();
});
