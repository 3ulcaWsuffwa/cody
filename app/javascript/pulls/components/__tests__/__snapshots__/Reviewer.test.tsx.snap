// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviewer snapshot test 1`] = `
<DocumentFragment>
  <li>
    <div
      class="flex items-center px-4 py-4 sm:px-6"
    >
      <div
        class="min-w-0 flex flex-1 items-center"
      >
        <div
          class="flex-shrink-0"
        >
          <span
            class="sr-only"
          >
            Approved
          </span>
          <svg
            class="flex-shrink-0 mr-1.5 h-5 w-5 text-green-500"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
            />
            <polyline
              points="22 4 12 14.01 9 11.01"
            />
          </svg>
        </div>
        <div
          class="min-w-0 flex-1 px-4 md:grid md:grid-cols-2 md:gap-4"
        >
          <div>
            <p
              class="text-sm font-medium truncate"
              data-testid="reviewer-login"
            >
              &lt;mock-value-for-field-"login"&gt;
            </p>
            <p
              class="text-sm truncate text-gray-500"
              data-testid="review-rule-name"
            >
              &lt;mock-value-for-field-"name"&gt;
            </p>
          </div>
        </div>
      </div>
    </div>
  </li>
</DocumentFragment>
`;
